tests/test_single_document.py

```diff
@@ -371,6 +346,54 @@ def test_progress(self) -> 'Generator':
         self.assertEqual(result, {"general": "kenobi"})
 
 
+class SingleDocumentTestCase2(TextDocumentTestCase):
+
+    def test_did_change(self) -> 'Generator':
+        assert self.view
+        self.maxDiff = None
+        self.insert_characters("A")
+        yield from self.await_message("textDocument/didChange")
+        # multiple changes are batched into one didChange notification
+        self.insert_characters("B\n")
+        self.insert_characters("🙂\n")
+        self.insert_characters("D")
+        promise = YieldPromise()
+        yield from self.await_message("textDocument/didChange", promise)
+        self.assertEqual(promise.result(), {
+            'contentChanges': [
+                {'rangeLength': 0, 'range': {'start': {'line': 0, 'character': 1}, 'end': {'line': 0, 'character': 1}}, 'text': 'B'},   # noqa
+                {'rangeLength': 0, 'range': {'start': {'line': 0, 'character': 2}, 'end': {'line': 0, 'character': 2}}, 'text': '\n'},  # noqa
+                {'rangeLength': 0, 'range': {'start': {'line': 1, 'character': 0}, 'end': {'line': 1, 'character': 0}}, 'text': '🙂'},  # noqa
+                # Note that this is character offset (2) is correct (UTF-16).
+                {'rangeLength': 0, 'range': {'start': {'line': 1, 'character': 2}, 'end': {'line': 1, 'character': 2}}, 'text': '\n'},  # noqa
+                {'rangeLength': 0, 'range': {'start': {'line': 2, 'character': 0}, 'end': {'line': 2, 'character': 0}}, 'text': 'D'}],  # noqa
+            'textDocument': {
+                'version': self.view.change_count(),
+                'uri': filename_to_uri(TEST_FILE_PATH)
+            }
+        })
+
+
+class SingleDocumentTestCase3(TextDocumentTestCase):
+
+    @classmethod
+    def get_test_name(cls) -> str:
+        return "testfile2"
+
+    def test_did_change_before_did_close(self) -> 'Generator':
+        assert self.view
+        self.view.window().run_command("chain", {
+            "commands": [
+                ["insert", {"characters": "TEST"}],
+                ["save", {"async": False}],
+                ["close", {}]
+            ]
+        })
+        yield from self.await_message('textDocument/didChange')
+        # yield from self.await_message('textDocument/didSave')  # TODO why is this not sent?
```

Comment:
now just to see why this happens
and than we can merge this

Reply:
test pass locally, but not on the CI
I am not sure what to do here, except to ignore this
and to report the issues to UnitTesting package.

Reply:
I can also see it happen locally, when I manually use the key binding in a file. Only `didChange` and `didClose` are sent, but not `didSave`. But this is another bug and not directly related to this PR, so we can investigate another time.

Reply:
When I uncomment the line.
and test the current test file, 
the tests pass on Linux and Mac locally:
```
test_apply_formatting (test_single_document.SingleDocumentTestCase) ... ok
test_apply_formatting_and_preserve_order (test_single_document.SingleDocumentTestCase) ... ok
test_declaration (test_single_document.SingleDocumentTestCase) ... ok
test_declaration_location_link (test_single_document.SingleDocumentTestCase) ... ok
test_definition (test_single_document.SingleDocumentTestCase) ... ok
test_definition_location_link (test_single_document.SingleDocumentTestCase) ... ok
test_did_close (test_single_document.SingleDocumentTestCase) ... ok
test_did_open (test_single_document.SingleDocumentTestCase) ... ok
test_expand_selection (test_single_document.SingleDocumentTestCase) ... ok
test_formats_on_save (test_single_document.SingleDocumentTestCase) ... ok
test_hover_info (test_single_document.SingleDocumentTestCase) ... ok
test_implementation (test_single_document.SingleDocumentTestCase) ... ok
test_implementation_location_link (test_single_document.SingleDocumentTestCase) ... ok
test_out_of_bounds_column_for_text_document_edit (test_single_document.SingleDocumentTestCase) ... ok
test_progress (test_single_document.SingleDocumentTestCase) ... ok
test_remove_line_and_then_insert_at_that_line_at_end (test_single_document.SingleDocumentTestCase) ... ok
test_rename (test_single_document.SingleDocumentTestCase) ... ok
test_run_command (test_single_document.SingleDocumentTestCase) ... ok
test_sends_save_with_purge (test_single_document.SingleDocumentTestCase) ... ok
test_tabs_are_respected_even_when_translate_tabs_to_spaces_is_set_to_true (test_single_document.SingleDocumentTestCase) ... ok
test_type_definition (test_single_document.SingleDocumentTestCase) ... ok
test_type_definition_location_link (test_single_document.SingleDocumentTestCase) ... ok
test_did_change (test_single_document.SingleDocumentTestCase2) ... ok
test_did_change_before_did_close (test_single_document.SingleDocumentTestCase3) ... ok
test_will_save_wait_until (test_single_document.WillSaveWaitUntilTestCase) ... ok

----------------------------------------------------------------------
Ran 25 tests in 6.032s

OK
```